viewing_party/party.py

```diff
@@ -0,0 +1,204 @@
+
+def get_watched_avg_rating(user_data):
+    num_movies = len(user_data["watched"])
+    if num_movies > 0:
+        sum = 0
+        for movie in user_data["watched"]:
+            sum += movie["rating"]
+        avg = sum / num_movies
+    else:
+        avg = 0.0
+    return avg
+
+
+def get_most_watched_genre(user_data): 
+    '''
+    determines which genre is most frequently occurring
+    return most frequently watched
+    if not applicable, return None
+    '''
+    # if there are watched movies, count genre freq
+    if len(user_data["watched"])> 0:
+        # store counters in genres
+        genres = dict()
+        for movie in user_data["watched"]:
+            if movie["genre"] in genres:
+                genres[movie["genre"]] += 1
+            else:
+                genres[movie["genre"]] = 1
+        # get max genre freq
+        return max(genres, key=genres.get) 
+    else:
+        return None
+
+def movie_list(person):
+    '''
+    return list of a users watched movies
+    '''
+    persons_movies = [movie["title"] for movie in person["watched"]]
+    return persons_movies
+
+def friends_watched_movies(user_data):
+    '''
+    return set of all friends movies combined
+    '''
+    friends_movie_list = []
+    for friend in user_data["friends"]:
+        friends_movie_list+=movie_list(friend)
+    friends_movie_set = set(friends_movie_list)
+    return friends_movie_set
+
+def get_unique_watched(user_data):
+    ''' 
+    return movies user has watched but no friend has
+    '''
+    # get users watched movies
+    user_movies = set(movie_list(user_data))
+    # get friends watched movies
+    friends_movie_set = friends_watched_movies(user_data)
+    # get difference between user and friends
+    unique_movies = user_movies - friends_movie_set
+    # store difference in dict
+    diff_dict = [{"title":movie} for movie in unique_movies]
+    return diff_dict
+
+def get_friends_unique_watched(user_data):
+    ''' 
+    return movies at least one of the user's 
+    friends have watched, but the user has not watched.
+    '''
+    # get movies watched by user
+    user_movies = set(movie_list(user_data))
+    # get movies watched by friends
+    friends_movie_set = friends_watched_movies(user_data)
+    # get diff between friends and user
+    unique_movies = friends_movie_set - user_movies 
+    # store diff in dict
+    diff_dict = [{"title":movie} for movie in unique_movies]
+    return diff_dict
+
+
+def get_available_recs(user_data):
+    ''' 
+    return list of recommended movies
+    recommended movies are not watched by user, 
+        a friend has watched it, & the user has the
+        subscription service
+    '''
+    # list of subscriptions
+    subs = user_data["subscriptions"]
+    # pair rec with service
+    movie_with_sub = []
+
+    for friend in user_data["friends"]:
+        for watched in friend["watched"]:   
+            if {"title":watched["title"]} in get_friends_unique_watched(user_data) and \
```

Comment:
this works great! we might be able to write a little less by saying:

```suggestion
            if watched in get_friends_unique_watched(user_data) and \
```

because dictionaries and lists are passed by reference, we can match an entire dictionary with one in a list

Reply:
Definitely agree! however the elements in friend["watched"] are in the form "{'title': 'Title A', 'host': 'Service A'}" whereas get_friends_unique_watched(user_data) is in the form {"title":watched["title]}. Now that I'm thinking about I suppose I could: 
1. check if a dictionary is a subset of another (generally seems more intensive than current method)
2. verify that just the key and value match (seems more intensive)
3. remove the host key from friend["watched"] and then check if its elements are in get_friends_unique_watched(user_data) (i would think removing 

Are there any other ways that could be better?

Reply:
ahh then that's definitely my mistake! I think your way would be more efficient than an alternative solution! thank you for explaining your thought process to me!